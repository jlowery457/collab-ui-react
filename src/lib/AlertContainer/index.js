import React from 'react';
import PropTypes from 'prop-types';
import Alert from '../Alert';
import {
  bind,
  reject,
  uniqueId
} from 'lodash';
import AlertCallContainer from '../AlertCallContainer';

class AlertContainer extends React.Component {
  static displayName = 'AlertContainer';

  state = {
    alertList: []
  }

  info = (title, message, onHide) => {
    this.handleAlert(
      title,
      message,
      onHide,
      'info'
    );
  }

  success = (title, message, onHide) => {
    this.handleAlert(
      title,
      message,
      onHide,
      'success'
    );
  }

  warning = (title, message, onHide) => {
    this.handleAlert(
      title,
      message,
      onHide,
      'warning'
    );
  }

  error = (title, message, onHide) => {
    this.handleAlert(
      title,
      message,
      onHide,
      'error'
    );
  }

  handleAlert = (title, message, onHide, type) => {
    const { orderNewest } = this.props;
    const key = uniqueId('alert_');
    const nextAlert = {
      key,
      title: title,
      message: message,
      onHide: bind(this.handleOnHide, this, key, onHide),
      type: type,
      show: true,
      closable: true
    }
    this.setState(state => ({
      alertList: orderNewest
        ? [nextAlert, ...state.alertList]
        : [...state.alertList, nextAlert]
    }));
  }

  handleOnHide = (key, onHide, e) => {
    this.setState(state => {
      onHide && onHide(e);
      return { alertList: reject(state.alertList, {key}) };
    })
  }

  render() {
    const { position } = this.props;
    return (
      <div className={`cui-alert__container cui-alert__container--${position}`}>
        {
          this.state.alertList.map(alert => (
            <Alert
              key={alert.key}
              title={alert.title}
              message={alert.message}
              type={alert.type}
              show={alert.show}
              closable={alert.closable}
              onHide={alert.onHide}
            />
          ))
        }
      </div>
    );
  }
}

AlertContainer.defaultProps = {
  orderNewest: true,
  position: 'bottom-right'
};

AlertContainer.propTypes = {
  /** Display new alert messages at the top or bottom of the queue */
  orderNewest: PropTypes.bool,
  /** Position alert will display */
  position: PropTypes.oneOf(['top-left', 'top-center', 'top-right', 'bottom-left', 'bottom-center', 'bottom-right'])
};

export default AlertContainer;
