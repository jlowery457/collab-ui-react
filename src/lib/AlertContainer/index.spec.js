import React from 'react';
import { shallow, mount } from 'enzyme';
import { Alert } from '@collab-ui/react';
import AlertContainer from '../AlertContainer';

describe('tests for <AlertContainer />', () => {
  const alertTitle = 'Now Hear This!';
  const alertMessage = 'Jefe Guadelupe, unit tesing like a boss!'

  it('should match SnapShot', () => {
    const container = shallow(<AlertContainer/>);

    expect(container).toMatchSnapshot();
  });

  it('should render a div in bottom-right by default', () => {
    const container = shallow(<AlertContainer />);

    expect(container.find('.cui-alert__container--bottom-right').length).toEqual(1);
  });

  it('should honor position prop when top-left is passed in', () => {
    const container = shallow(<AlertContainer position={'top-left'}/>);

    expect(container.find('.cui-alert__container--top-left').length).toEqual(1);
  });

  it('should honor position prop when top-center is passed in', () => {
    const container = shallow(<AlertContainer position={'top-center'}/>);

    expect(container.find('.cui-alert__container--top-center').length).toEqual(1);
  });

  it('should honor position prop when top-right is passed in', () => {
    const container = shallow(<AlertContainer position={'top-right'}/>);

    expect(container.find('.cui-alert__container--top-right').length).toEqual(1);
  });

  it('should honor position prop when bottom-left is passed in', () => {
    const container = shallow(<AlertContainer position={'bottom-left'}/>);

    expect(container.find('.cui-alert__container--bottom-left').length).toEqual(1);
  });

  it('should honor position prop when bottom-center is passed in', () => {
    const container = shallow(<AlertContainer position={'bottom-center'}/>);

    expect(container.find('.cui-alert__container--bottom-center').length).toEqual(1);
  });

  it('should honor position prop when bottom-right is passed in', () => {
    const container = shallow(<AlertContainer position={'bottom-right'}/>);

    expect(container.find('.cui-alert__container--bottom-right').length).toEqual(1);
  });

  it('should render an info Alert when info() is called', () => {
    const container = mount(<AlertContainer/>);
    container.instance().info(
      alertTitle,
      alertMessage,
      () => {}
    );
    container.update();
    expect(container.find('.cui-alert--info').length).toEqual(1);
  });

  it('should render a success Alert when success() is called', () => {
    const container = mount(<AlertContainer/>);
    container.instance().success(
      alertTitle,
      alertMessage,
      () => {}
    );
    container.update();
    expect(container.find('.cui-alert--success').length).toEqual(1);
  });

  it('should render a warning Alert when warning() is called', () => {
    const container = mount(<AlertContainer/>);
    container.instance().warning(
      alertTitle,
      alertMessage,
      () => {}
    );
    container.update();
    expect(container.find('.cui-alert--warning').length).toEqual(1);
  });

  it('should render an error Alert when error() is called', () => {
    const container = mount(<AlertContainer/>);
    container.instance().error(
      alertTitle,
      alertMessage,
      () => {}
    );
    container.update();
    expect(container.find('.cui-alert--error').length).toEqual(1);
  });

  it('number of Alerts should equal number of times functions are called', () => {
    const container = mount(<AlertContainer/>);
    const alertCount = 5;

    for (let index = 0; index < alertCount; index++) {
      container.instance().info(
        alertTitle,
        alertMessage,
        () => {}
      );
    }
    container.update();
    expect(container.find(Alert).length).toEqual(alertCount);
  });
});
