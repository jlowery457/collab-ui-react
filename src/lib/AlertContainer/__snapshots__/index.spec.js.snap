// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tests for <AlertContainer /> should match SnapShot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <AlertContainer
    newestOnTop={true}
    position="bottom-right"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [],
      "className": "cui-alert__container cui-alert__container--bottom-right",
    },
    "ref": null,
    "rendered": Array [],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [],
        "className": "cui-alert__container cui-alert__container--bottom-right",
      },
      "ref": null,
      "rendered": Array [],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactFifteenAdapter {
      "options": Object {
        "supportPrevContextArgumentOfComponentDidUpdate": true,
      },
    },
  },
}
`;
